pageserver/src/http/routes.rs

```diff
@@ -224,22 +224,21 @@ async fn timeline_detail_handler(request: Request<Body>) -> Result<Response<Body
             info_span!("timeline_detail_handler", tenant = %tenant_id, timeline = %timeline_id)
                 .entered();
         let repo = tenant_mgr::get_repository_for_tenant(tenant_id)?;
-        Ok::<_, anyhow::Error>(match repo.get_timeline(timeline_id)?.local_timeline() {
-            None => TimelineInfo::Remote {
-                timeline_id,
-                tenant_id,
-            },
-            Some(timeline) => TimelineInfo::Local {
+        let local_timeline = {
+            // TODO (current) handle other timeline states
+            let timeline = repo.get_local_timeline_load(timeline_id)?;
```

Comment:
So that will implicitly load the timeline into pageserver's memory.

Curious a bit: where we draw the line between loading into the memory and S3 downloading?
`zenith init` timeline state can be downloaded and loaded into memory in a matter of a couple of seconds now, so from my perspective, both download and load into memory could be on par speed-wise, so we could in theory treat them as different stages of the timeline loading and require both to be done explicitly?

Reply:
To paraphrase: right now, it looks like we still have some implicitness on how we treat the not-loaded-in-memory layers, mainly we seem to load such layers implicitly as here. Feels odd, compared to what we've just done to separate the download into a separate manual step.

Reply:
> could be on par speed-wise

I think this highly depends on how our storage is organized and what is the size of the timeline. For large databases and current storage design with many relatively small layers loading of the layer map can take an hour to complete. But I think with upcoming LSM approach loading will become significantly faster.

> to load such layers implicitly as here

See TODO above :) Timeline detail should be free of side effects.

Regarding different stages of load, I have couple of open questions which require additional investigation. In general for timeline to be ready we need both, locally available data and basic index in memory. 
* How to match SLO on new compute start? Do we always keep timeline assigned to at least one pageserver or we can completely unload it. In that case cold start can take quite a long time especially for large timelines. IMO we should keep timeline on at least one pageserver
* In the upcoming storage design load from disk should not take too much time so I expect loading from s3 to take more time (but for now this is just speculation). I think we can require timeline to be loaded from disk only in the places it is actually needed (e.g. when compute connected to us or we've received callmemaybe request). And this PR opens possibility to do so. Returning an error in the above cases when timeline is not available locally is ok to me because it is actually an error to direct traffic to the pageserver which is not ready to accept it. This error indicates that we cannot start processing in a reasonable time (i e without violating our SLO). That may be the reason why we need to treat s3 sync differently from load from disk.
* Can we/should we interleave download with loading? I e download some index parts, and begin bootstrapping it before all the data is available locally. It might be quite hard to organize given our mix of sync/async parts of repository/remote storage sync. 

It is still an open question how to organize it better (especially for our future demands) and answer all the questions at once. But anyway we should start to evolve it in some promising direction